R/utils.R

```diff
@@ -485,15 +472,122 @@ multi_choice_to_labels <- function(db_data, db_metadata) {
-        as.character() %>%
-        recode(!!!parse_labels_output) %>%
-        factor(levels = unique(parse_labels_output))
+
+      db_data[[field_name]] <- label_handler(
+        x = db_data[[field_name]],
+        labels = parse_labels_output,
+        ptype = db_data[[field_name]]
+      )
     }
   }
   db_data
 }
 
+#' @title
+#' Apply factor labels to a vector
+#'
+#' @param x a vector to label
+#' @param labels a named vector of labels in the format `c(value = label)`
+#' @param \dots unused
+#'
+#' @return
+#' factor
+#'
+#' @keywords internal
+apply_labs_factor <- function(x, labels, ...) {
+  as.character(x) %>%
+    recode(!!!labels) %>%
+    factor(levels = unique(labels))
+}
+
+#' @title
+#' Apply haven value labels to a vector
+#'
+#' @details
+#' Assumes a check_installed() has been run for `labelled`. Since `haven` preserves the
+#' underlying data values we need to make sure the data type of the value options in the metadata matches
+#' the data type of the values in the actual data. This function accepts a prototype, usually a column
+#' from db_data, and uses `force_cast()` to do a best-effort casting of the value options in the metadata
+#' to the same data type as `ptype`. The fallback is to convert `x` and the value labels to character.
+#'
+#' @param x a vector to label
+#' @param labels a named vector of labels in the format `c(value = label)`
+#' @param ptype vector to serve as prototype for label values
+#' @param \dots unused
+#'
+#' @return
+#' `haven_labelled` vector
+#'
+#' @keywords internal
+#'
+apply_labs_haven <- function(x, labels, ptype, ...) {
+  # set_value_labels expects labels in c(label = value) format so reverse them
+  labels <- invert_vec(labels)
+  # Try to cast values to match data type in data, catching any parsing warnings
+  cnd <- NULL
+  labels_cast <- withCallingHandlers(
+    {
+      force_cast(labels, ptype)
+    },
+    warning = function(w) {
+      cnd <<- w
+      cnd_muffle(w)
+    }
+  )
+  if (!is.null(attr(labels_cast, "problems"))) {
+    # If there was parsing problem fall back to character
+    x <- as.character(x)
+    labels_cast <- force_cast(labels, character())
+  } else if (!is.null(cnd)) {
+    # If there was some other warning we didn't mean to catch it, so re-raise
+    cli_warn(cnd)
+  }
+
+  labelled::set_value_labels(x, .labels = labels_cast)
+}
+
+#' @title
+#' Swap vector names for values
+#'
+#' @param x a vector
+#'
+#' @return
+#' Vector with names and values reversed
+#'
+#' @keywords internal
+#'
+invert_vec <- function(x) {
+  out <- names(x)
+  # If there were no names do nothing
+  if (is.null(out)) {
+    return(x)
+  }
+  names(out) <- x
+  out
+}
+
+force_cast <- function(x, ptype) {
+  ptype <- vec_ptype(ptype)
+  if (is.logical(ptype)) {
+    out <- parse_logical(x)
+  } else if (is.integer(ptype)) {
+    out <- parse_integer(x)
+  } else if (is.numeric(ptype)) {
+    out <- parse_double(x)
+  } else if (is.Date(ptype)) {
+    out <- parse_date(x)
```

Comment:
I think we can address these Codecov warnings by adding the associated data types to tests in your new tests. I don't think it's useful to write a test for this function in particular since it's so basic, but you could do either.

https://github.com/CHOP-CGTInformatics/REDCapTidieR/blob/c58e78363fa246589a0bbfd1d05554c1164f9e01/tests/testthat/test-utils.R#L436

Reply:
Good call. Will do! Btw are those code coverage warnings getting added to the conversation new? I've never seen them before

Reply:
Unsure, I started noticing them in the past month or two. Relatively helpful for helping keep our % up

Reply:
Okay interesting finding (and I think why I skipped including those tests to start).

With `labelled` if you try to apply labels to vector that has no corresponding `val_labels<-` method it just returns the unlabelled vector:

```r
labels <- as.Date(c("The first day of 2023" = "2023-01-01"))

x <- labelled::set_value_labels(as.Date("2023-01-01"), .labels = labels)

x
#> [1] "2023-01-01:

class(x)
#> [1] "Date"
```

That means if someone has some crazy redcap where the value options are dates and they do `raw_or_label = "haven"`, their value labels won't get applied at all. BUT if they went ahead and implemented a `val_labels<-.Date` method then they would get applied. This is such a weird situation I don't think we should even care tbh. I added tests for `force_cast()` just to get rid of the coverage warnings.

Reply:
Huh, very interesting find. In that case yea, much better to just test `force_cast()`. If for _some_ reason someone encounters this then an issue can be opened. 

I figured we wouldn't test `force_cast()` since most of the options for it were already tested in the parent function but whatever appeases the CodeCov gods is fine.

Reply:
Value labels could be applied only to character or numeric vectors. It is not possible to add value labels to date vectors.

In fact, an error should have occured. I have open an issue on labelled website: https://github.com/larmarange/labelled/issues/156